File: client/Ciphelet/app/src/main/java/com/github/worldoneo/ciphelet/connector/ConnectorThread.java
package com.github.worldoneo.ciphelet.connector;

import android.content.SharedPreferences;
import android.util.Base64;

import androidx.core.util.Consumer;

import com.github.worldoneo.ciphelet.MainActivity;
import com.github.worldoneo.ciphelet.R;
import com.github.worldoneo.ciphelet.connector.encryption.EncryptionUtility;
import com.github.worldoneo.ciphelet.connector.events.EventManager;
import com.github.worldoneo.ciphelet.connector.events.LoginSuccessEvent;
import com.github.worldoneo.ciphelet.storage.SecureStorage;

import java.net.URI;
import java.net.URISyntaxException;

public class ConnectorThread extends Thread {
    private final String password;
    private final URI u;
    private final SharedPreferences preferences;
    private SecureStorage secureStorage;


    public ConnectorThread(String password, URI u, SharedPreferences preferences) {
        this.password = password;
        this.u = u;
        this.preferences = preferences;
    }

    @Override
    public void run() {
        this.secureStorage = new SecureStorage(preferences, EncryptionUtility.getKeyFromPassword(password, SecureStorage.getSalt(preferences)));
        System.out.println("Going async!");
        URI uri = null;
        try {
            uri = new URI(MainActivity.getInstance().getStringsxml(R.string.server));
        } catch (URISyntaxException e) {
            e.printStackTrace();
        }

        CipheletAPI cipheletAPI = new CipheletAPI(
                secureStorage.get("userid", Long.class),
                new Connector(uri),
                EncryptionUtility.decodeKey(secureStorage.get("privatekey", String.class)));
        cipheletAPI.onLogin(new Consumer<CipheletAPI>() {
            @Override
            public void accept(CipheletAPI cipheletAPI) {
                EventManager.getInstance().handleEvent(new LoginSuccessEvent(cipheletAPI));
            }
        });
        cipheletAPI.login(password);
    }
}

File: client/Ciphelet/app/src/main/java/com/github/worldoneo/ciphelet/connector/ConnectorThread.kt
package com.github.worldoneo.ciphelet.connector

import android.content.SharedPreferences
import androidx.core.util.Consumer
import com.github.worldoneo.ciphelet.MainActivity
import com.github.worldoneo.ciphelet.R
import com.github.worldoneo.ciphelet.connector.encryption.EncryptionUtility
import com.github.worldoneo.ciphelet.connector.events.EventManager.handleEvent
import com.github.worldoneo.ciphelet.connector.events.LoginSuccessEvent
import com.github.worldoneo.ciphelet.storage.SecureStorage
import java.lang.RuntimeException
import java.net.URI
import java.net.URISyntaxException

class ConnectorThread(private val password: String, private val u: URI, private val preferences: SharedPreferences) : Thread() {
    private var secureStorage: SecureStorage? = null
    override fun run() {
        val key = EncryptionUtility.getKeyFromPassword(password, SecureStorage.getSalt(preferences))
                ?: throw RuntimeException()
        secureStorage = SecureStorage(preferences, key)
        println("Going async!")
        var uri: URI? = null
        try {
            uri = URI(MainActivity.instance!!.getStringsxml(R.string.server))
        } catch (e: URISyntaxException) {
            e.printStackTrace()
        }
        val userid = secureStorage!!.get("userid", Long::class.java) ?: throw RuntimeException()
        val cipheletAPI = CipheletAPI(
                userid,
                Connector(uri),
                EncryptionUtility.decodeKey(secureStorage!!.get("privatekey", String::class.java)))
        cipheletAPI.onLogin(Consumer { handleEvent(LoginSuccessEvent(cipheletAPI)) })
        cipheletAPI.login(password)
    }

}